src/task4feedback/simulator/rl/models/simple.py

```diff
@@ -0,0 +1,295 @@
+        # S, (S, A) value networks
+        self.network = A2CNetworkNoGCN(rl_env.get_state_dim(), rl_env.get_out_dim())
+        self.optimizer = optim.RMSprop(self.network.parameters(),
+                                       lr=0.0001)
+                                       #lr=0.0005)
+        self.execution_mode = execution_mode
+        self.is_loaded_model_best = load_best_model
+        self.fastest_execution_time = float("inf")
+        self.lr = lr
+        self.episode = 0
+        self.a2cnet_fname = "a2c_network.pt"
+        self.optimizer_fname = "optimizer.pt"
+        self.best_a2cnet_fname = "best_a2c_network.pt"
+        self.best_optimizer_fname = "best_optimizer.pt"
+        self.device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+        # Accumulated reward 
+        self.accumulated_reward = 0
+
+        # pi(state) -> action probabilities
+        self.pi = {}
+        # lambda to update a target policy
+        self.ld = 0.5
+
+        # Buffer (S, A, S') until the terminal state
+        self.tmp_curr_state = None
+        self.tmp_next_state = None
+        self.tmp_action = None
+        self.logs: List[MappingLogs] = []
+
+        if not self.is_training_mode():
+            print("Load model..\n")
+            if self.is_loaded_model_best == 1:
+                self.load_best_model()
+            else:
+                self.load_model()
+
+    def select_device(self, state: torch.tensor, oracle):
+        """
+        Select a device from pi.
+        If a specified state has not been visited, select a device from a neural
+        network.
+        """
+
+        state_tuple = tuple(state.tolist())
+        with torch.no_grad():
+            if state_tuple in self.pi:
+                # If the current state is already in pi, sample an action from
+                # the stored action probabilities
+                dist = Categorical(self.pi[state_tuple])
+                action = dist.sample()
+                # print("State existed: ", action)
+            else:
+                actions, v = self.network(NetworkInput(state, False, None, None))
+                self.pi[state_tuple] = F.softmax(actions, dim=0)
+                dist = Categorical(self.pi[state_tuple])
+                action = dist.sample()
+                # print("action: ", action, " pi:", self.pi[state_tuple])
+            if self.is_training_mode():
+                self.update_pi(state, state_tuple, oracle)
+        return action
+
+    def update_pi(self, state: torch.tensor, state_tuple: Tuple[float], oracle):
+        """ Update target policy, pi.
+        """
+
+        # Buffer probabilities from network
+        actions, v = self.network(NetworkInput(state, False, None, None))
+        actions = F.softmax(actions, dim=0)
+        # print("oracle:", oracle, " softmax dnn:", actions)
+        # Calculate loss
+        self.pi[state_tuple] = (1 - self.ld) * actions + self.ld * oracle.to(self.device)
```

Comment:
I think ld should be random. 
We discussed it as a bernoulli rv to toggle between either at each step

Reply:
Although (if training is fast), it is maybe worth seeing if this converges even with a static ld.